Add Julia port of Openlibm asin

pkofod opened the pull request:
Based on openlibm, so I should do the whole license-thing
 - [x] fix license info

`asin` is not really tested in Base besides some basic tests that it works, as it is assumed that openlibm tests itself, so I might want to add some tests here. 
- [x] add tests

```julia

using Plots
plotly()
function scaled_big_error{T}(f, x::T)
    fx = f(x)
    fbx = f(big(x))
    abs(big(fx)-fbx)/big(eps(T(fbx)))
end

histogram(linspace(-1,1, 1000000), x->scaled_big_error(asin, x))
```

Should be equivalent to openlibm. Accuracy is often within 0.5 ulp, but can go higher.
*Float64*
![asin](https://user-images.githubusercontent.com/8431156/28864450-dfcdaea4-776c-11e7-9a70-2ba4bfcaed6b.png)
*Float32*
<s>(just realized that those computations should be in Float32 as well... will fix)</s>
![asinf](https://user-images.githubusercontent.com/8431156/28921423-33388bf6-7856-11e7-941b-0c62b0a62db2.png)

Tests ran succesfully locally.

*Common `asin` function with type specific kernels*
I've merged the two methods in one main methods calling out to `@horner` based methods for the polynomials `p` and `q`.

pkofod committed:
Add Julia implementation of asin.

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +76,129 @@ function cos_kernel(x::DoubleFloat32)
+    pS0 =  1.6666586697e-01
+    pS1 = -4.2743422091e-02
+    pS2 = -8.6563630030e-03
+    qS1 = -7.0662963390e-01
```
Is this really supposed to evaluated in Float64 ?

pkofod replied:
No, see the comment just above the histogram :)

musm reviewed (commented):
Looks like both version of FLoat32 and Float64, have a lot of duplicated code which could be merged into one method dispatching to custom kernels

pkofod commented:
something like this @musm?

KristofferC commented on base/special/trig.jl:
```diff
@@ -76,6 +76,115 @@ function cos_kernel(x::DoubleFloat32)
+    pio2_lo =  6.12323399573676603587e-17
+    s = sqrt(t)
+
+    if absx >= 0.975 # if |x| > 0.975 */
```
`absx` not defined.

pkofod replied:
`absx` police? :) Thanks

KristofferC commented on base/special/trig.jl:
```diff
@@ -76,6 +76,109 @@ function cos_kernel(x::DoubleFloat32)
+    end
+    # else 1/2 <= |x| < 1.
+    t = (one(T)-absx)/2
+    return flipsign(asin_Rx(p, q, t), x)
```
`p` police

tkelman commented on base/special/trig.jl:
```diff
@@ -76,6 +76,109 @@ function cos_kernel(x::DoubleFloat32)
+    pS3 = -4.00555345006794114027e-02
+    pS4 =  7.91534994289814532176e-04
+    pS5 =  3.47933107596021167570e-05
+    t*(pS0+t*(pS1+t*(pS2+t*(pS3+t*(pS4+t*pS5)))))
```
isn't this what `@horner` is for?

pkofod replied:
It is indeed

maleadt commented on base/special/trig.jl:
```diff
@@ -76,6 +76,105 @@ function cos_kernel(x::DoubleFloat32)
+    flipsign(Float32(pi/2-2*(s+s*Rx)), x)
+end
+
+function asin{T<:AbstractFloat}(x::T)
```
deprecated parametric function syntax, needs `where`

pkofod replied:
alright, need to read up on that then :)

KristofferC commented:
For `asin(0.54)`:

* Moving the `throw` statement to a separate `@noinline` function : 14.8ns to 13.1ns.

* Forcing inlining of `asin_kernel`, 13.1ns to 12.7ns.

pkofod commented:
> For (asin(0.54), asin(0.54f0)):
> 
>     Moving the throw statement to a separate @noinline function : (12.65ns, 8.5ns) to (11.5ns, 7.4ns).
>     Forcing inlining of asin_kernel, (11.5ns, 7.4ns) to (11.2ns, 6.65ns).
> 
> Base runs at (16.1 ns, 9.6 ns) for me.
> 
> Edit: Benchmarks were on 0.6, probably different on 0.7. Updated with 0.7

Seems consistent with my experience from `rem_pio2` and other functions. I just wanted the refactoring to work first! :)

Good to see speed improvements. I actually added benchmarks to BaseBenchmarks (https://github.com/JuliaCI/BaseBenchmarks.jl/pull/87) but I don't think it's been retuned since. Anyway, shouldn't run those before inlining.

KristofferC commented on base/special/trig.jl:
```diff
@@ -158,7 +159,7 @@ function asin{T<:AbstractFloat}(x::T)
             return flipsign(T(pi)/2, x)
         end
-        throw(DomainError(x, "asin(x) is not defined for |x|>1."))
+        noinline_domainerror(x, "asin(x) is not defined for |x|>1.")
```
Move the string to the non inlined function.

pkofod replied:
is this style or performance? I'm going to do the same in other functions and it seems a bit odd to have one for `acos` as well, but if it's needed for performance, I'll do it ofc :)

KristofferC replied:
Performance, gotta hide them strings to the GC.

pkofod commented:
Alright, when tuning is done, a nanosoldier run would be sweet here @ararslan  (since I figure you will  know when it's tuned)

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +76,105 @@ function cos_kernel(x::DoubleFloat32)
+        # tan(x) ≈ x+x*R(x)
+        x² = x*x
+        Rx = asin_p(x²)/asin_q(x²) # rational approximation
+        return x+x*Rx
```
seems like this could  benefit from `muladd`

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +76,105 @@ function cos_kernel(x::DoubleFloat32)
+                             Float32(-4.2743422091e-02),
+                             Float32(-8.6563630030e-03))
+
+asin_q(t::Float32) = 1.0f0+t*Float32(-7.0662963390e-01)
```
Haven't had a change to look at things in detail but just use the float syntax here and above, e.g.
`-7.0662963390f-01`

seems like this could also benefit from `muladd`

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +76,105 @@ function cos_kernel(x::DoubleFloat32)
+        Rx0 = reinterpret(Float64, (reinterpret(UInt64, s) >> 32) << 32)
+        c = (t-Rx0*Rx0)/(s+Rx0)
+        r = p/q
+        p = 2.0*s*r-(pio2_lo-2.0*c)
```
can I request spaces?
`2*s*r - (pio2_lo - 2*c)`

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +76,105 @@ function cos_kernel(x::DoubleFloat32)
+end
+
+@noinline asin_domain_error(x) = throw(DomainError(x, "asin(x) is not defined for |x|>1."))
+function asin{T<:AbstractFloat}(x::T)
```
This should be restricted to `Union{Float32,Float64}`

tkelman replied:
and isn't this syntax deprecated?

musm commented on base/math.jl:
```diff
@@ -436,7 +436,7 @@ julia> log1p(0)
+for f in (:sin, :cos, :tan, :acos, :acosh, :atanh, :log, :log2, :log10,
           :lgamma, :log1p)
     @eval begin
         @inline ($f)(x::Float64) = nan_dom_err(ccall(($(string(f)), libm), Float64, (Float64,), x), x)
```
unrelated by I wonder why these have inline calls

pkofod committed:
Add asin tests.

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +76,105 @@ function cos_kernel(x::DoubleFloat32)
+    end
+end
+@inline function asin_kernel(t::Float32, x)
+    s = sqrt(Float64(t))
```
This should not have `Float64`

pkofod replied:
Are you sure? As far as I can see s is the only `double` variable in `asinf`, and it holds the returned value from `sqrt(t)` not `sqrtf(t)`. I'm no C wiz though, so please do correct me if I misunderstand the code!

https://github.com/JuliaLang/openlibm/blob/0598080ca09468490a13ae393ba17d8620c1b201/src/e_asinf.c

ararslan commented on test/math.jl:
```diff
@@ -678,3 +678,26 @@ let x = 2.0
+        @test asin(0.6f0) == 0.64350116f0
+        @test asin(-0.6f0) == -0.64350116f0
+        @test asin(0.98f0) == 1.3704616f0
+        @test asin(-0.98f0) == -1.3704616f0
```
Should unindent these since they're outside of the block above

jebej commented on base/special/trig.jl:
```diff
@@ -76,6 +76,105 @@ function cos_kernel(x::DoubleFloat32)
+        Rx = p/q
+        return flipsign(pi/2 - (2.0*(s + s*Rx) - pio2_lo), x)
+    else
+        Rx0 = reinterpret(Float64, (reinterpret(UInt64, s) >> 32) << 32)
```
I'm curious about this line, what is it doing?

pkofod replied:
It's zeroing out the low word of `s`. I changed the name of the variables here.

jebej commented on base/special/trig.jl:
```diff
@@ -76,6 +76,105 @@ function cos_kernel(x::DoubleFloat32)
+    p = asin_p(t) # numerator polynomial
+    q = asin_q(t) # denominator polynomial
+    Rx = p/q # rational approximation
+    flipsign(Float32(pi/2 - 2*(s + s*Rx)), x)
```
should the Float32 just be on pi here?

pkofod replied:
I may have misunderstood the original code, but since `pio2` is specifically `double`, and `2.0` is not `(float)2.0` it was my understanding that https://github.com/JuliaLang/openlibm/blob/0598080ca09468490a13ae393ba17d8620c1b201/src/e_asinf.c#L64 was doing what is effectively extended precision calculation and then converting back to `float`.

jebej replied:
I suppose that was just in case the sqrt above was supposed to be a Float32, but looking at the C code it seems you are consistent with it. It does seem strange for `s` to be a double in the C code, since they are taking the sqrt of a float. I guess that last digit will benefit from the better pi approximation. Wonder how much that improves accuracy...

pkofod replied:
Well, since this is floats, I guess we can just check!

tkelman commented on test/math.jl:
```diff
@@ -678,3 +678,20 @@ let x = 2.0
+
+@testset "asin #23088" begin
+    for T in (Float32, Float64)
+        @test asin(zero(T)) == zero(T)
```
`===` to check return type as well?

pkofod replied:
Good idea, will change it later

giordano commented on test/math.jl:
```diff
@@ -678,3 +678,20 @@ let x = 2.0
+        @test asin(one(T)) == T(pi)/2
+        @test asin(-one(T)) == -T(pi)/2
+        for x in (0.45, 0.6, 0.98)
+            by = T(asin(big(x)))
```
Strictly speaking, the conversion to `BigFloat` in this way is unsafe, the best way to initialize a `BigFloat` is by using strings.  E.g.:
```julia
julia> big(0.45)
4,500000000000000111022302462515654042363166809082031250000000000000000000000000e-01

julia> big"0.45"
4,499999999999999999999999999999999999999999999999999999999999999999999999999983e-01

julia> Float64(asin(big(0.45)))
0.4667653390472964

julia> Float64(asin(big"0.45"))
0.46676533904729633
```
Anyway, for the precision required this may not be an issue.

pkofod replied:
Good to know, thanks. I just followed the style from earlier in this file

giordano replied:
If you add 0.9999 to the list of numbers to test, the current way to test would report that the error is zero, while it's actually 4 ulp if you use the correct way to convert to `BigFloat`:
```julia
julia> x = 0.9999
0.9999

julia> by = Float64(asin(big(x)))
1.5566540733173846

julia> asin(x)
1.5566540733173846

julia> abs(asin(x) - by) / eps(by)
0.0

julia> By = Float64(asin(parse(BigFloat, repr(x))))
1.5566540733173837

julia> abs(asin(x) - By) / eps(by)
4.0
```
The error growths as `x` approaches 1, up to 3.418348e6 ulp for `x = 0.9999999999999999`.

However,
```julia
julia> sin(1.5566540733173846)
0.9999

julia> sin(1.5566540733173837)
0.9999
```
so both numbers are mapped to `0.9999` by `sin`.

It looks like that an accurate docstring of `asin` would be

> Compute a number that is mapped into `x` by `sin`

rather than

> Compute the inverse sine of `x`

This has nothing to do with the proposed implementation of the function, also the previous one gives the same results.

pkofod commented:
I'm condensed it to two commits. First, tests are added that runs on master, and then the julia version is added.

ararslan commented:
FreeBSD failure is #23143.

@nanosoldier `runbenchmarks(ALL, vs=":master")`

yuyichao commented on base/special/trig.jl:
```diff
@@ -76,6 +76,105 @@ function cos_kernel(x::DoubleFloat32)
+
+@inline function asin_kernel(t::Float64, x)
+    pio2_lo = 6.12323399573676603587e-17
+    s = sqrt(t)
```
Is llvm smart enough to remove the `t < 0` branch?

pkofod replied:
Don't know, are you just curious, or is there something like a `pos_sqrt(t)` function?

yuyichao replied:
There's `FastMath.sqrt_fast` and `sqrt_llvm`.

pkofod replied:
Is `FastMath.sqrt_fast` just a mapping to `sqrt_llvm`? And if so, where can I find the documentation for `sqrt_llvm`?

yuyichao commented on test/math.jl:
```diff
@@ -678,3 +678,20 @@ let x = 2.0
+            by = T(asin(big(x)))
+            @test abs(asin(T(x)) - by)/eps(by) <= one(T)
+            bym = T(asin(big(-x)))
+            @test abs(asin(T(-x)) - bym)/eps(bym) <= one(T)
```
Test for `NaN` and other non-finite numbers?

yuyichao reviewed (changes requested):
You should probably also update this for fastmath version.

ararslan commented:
Had to give the Nanosoldier a whack. Let's try this again.

@nanosoldier `runbenchmarks(ALL, vs=":master")`

tkelman commented:
> - [x] fix license info

I don't see any license related changes here, should this be checked yet?

pkofod commented:
> I don't see any license related changes here, should this be checked yet?

Forgot to add the change before committing and pushing.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23088#issuecomment-320467940) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/72576b7a0b90f42007ffae0ef4d3fc6090ed9ffa/a4daf3b_vs_83007fb/report.md). cc @ararslan

pkofod commented:
> Your benchmark job has completed - possible performance regressions were detected. A full report can be found here. cc @ararslan

With fixed layout
https://gist.github.com/pkofod/13076ea3e55775fea4e75880bf025a5b

ararslan commented:
Impressive improvements in the asin benchmarks!

pkofod commented:
The requested changes by @yuyichao has a red cross but I believe I have fixed them, it just didn't hit the appropriate line!

tkelman commented on base/special/trig.jl:
```diff
@@ -105,7 +105,7 @@ asin_q(t::Float32) = muladd(t, -7.0662963390f-01, 1.0f0)
 @inline function asin_kernel(t::Float64, x)
     pio2_lo = 6.12323399573676603587e-17
-    s = sqrt(t)
+    s = sqrt_llvm(t)
```
what's the difference?

musm replied:
This avoid's  the unnecessary error check for t < 0, so it's faster.

martinholters commented:
What is going on with those four broadcast benchmarks?

pkofod commented:
> What is going on with those four broadcast benchmarks?

Don't know. Don't think `asin` should be involved in addition, subtraction or exponentiation of `Float64`'s.

pkofod commented:
I think it's done, but comments are still welcome. I did make some small adjustments since last nanosoldier run, so feel free to run again. Don't know if you have anything to say @simonbyrne , but it seems like you were requested for review, don't know how official that request is.

StefanKarpinski commented:
It's always nice to get an approval from @simonbyrne since he's very good at this sort of code, but I don't think it's necessary. You've done accuracy comparisons for 64 and 32 bit. What I would consider doing myself to be sure this gets all corner cases is do a comprehensive sweep on Float32 since that's feasible for a single-argument function. If you did that already, then we should merge.

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +77,105 @@ function cos_kernel(x::DoubleFloat32)
+end
+
+@noinline asin_domain_error(x) = throw(DomainError(x, "asin(x) is not defined for |x|>1."))
+function asin{T<:Union{Float32, Float64}}(x::T)
```
update the parametric syntax here

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +77,105 @@ function cos_kernel(x::DoubleFloat32)
+    flipsign(Float32(pi/2 - 2*(s + s*Rx)), x)
+end
+
+@noinline asin_domain_error(x) = throw(DomainError(x, "asin(x) is not defined for |x|>1."))
```
I'd just remove this and put the `throw(DomainError(x, "asin(x) is not defined for |x|>1."))` directly in the method.
Also may be nice to do `|x| > 1` (spaced out)  or `abs(x) >1`

fredrikekre replied:
https://github.com/JuliaLang/julia/pull/23088#issuecomment-320279238

musm replied:
Interesting. That is pretty annoying, I wonder if there is a heuristic that can be  tuned so that this isn't always an issue ? I am sure I am not the only one who could potentially be bit by this.

musm replied:
I am unable to benchmark this on master due to a bug in `BenchmarkTools`

musm replied:
At least on 0.6,  I was able to benchmark this and noted no major differences (with the inline `DomainError` being slightly better in some circumstances, but only very minor) .

pkofod replied:
Yeah, kind of annoying, but several macros seem to not function on master... If it has no impact I'd be more than happy to not define a new function to do this for each elementary function that I'll submit PRs for, but @KristofferC 's benchmarks seemed to show somewhat significant impact of not doing it this way.

KristofferC replied:
Moving the throw statement made a significant impact for and the generated code is much smaller.

musm replied:
How were u able to benchmark on master ? The length of the generated code shouldn't affect the timing performance in this case, but I'm not 100% sure.  Maybe @timholy could provide some insight.

KristofferC replied:
Compare e.g.:

```
function f(x)
    if x < 0
        throw(ArgumentError("String"))
    else
        return x
    end
end
```

vs

```
function f2(x)
    if x < 0
        throw_err()
    else
        return x
    end
end

@noinline throw_err() = throw(ArgumentError("String"))
```

musm replied:
I am not sure if this is a good comparison since there is no computation being performed. This would suggest *all* uses of  `throw(...)` should use an  outer function, which I suspect should not be be the case. 

For example, I benchmarked the `asin` function for the two alternatives (inline `throw` and outer method with `@noinline` for the `throw`) on 0.6  and noted no difference.  Might have to wait until `BenchmarkTools` is fixed to benchmark master.

KristofferC replied:
I don't see a difference now on the laptop but the code has changed since I benchmarked so not sure if something else had an impact.

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +77,105 @@ function cos_kernel(x::DoubleFloat32)
+                             -4.2743422091f-02,
+                             -8.6563630030f-03)
+
+asin_q(t::Float32) = muladd(t, -7.0662963390f-01, 1.0f0)
```
Maybe write this as `@horner( t, 1f0, -7.0662963390f-01)` so it parallels the above code.

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +77,105 @@ function cos_kernel(x::DoubleFloat32)
+    # Special cases:
+    #    if |x|>1, throw DomainError
+    absx = abs(x)
+    if absx >= one(T) # |x|>= 1
```
In the other math function, I elected to use `T(1.0)` instead of `one(T)`, since the functions are only defined for Float32 and Float64, and since I also needed other numeric constants, e.g, `T(2.0)` (and we don't have `two(T)` 😉 ). Just thought I would mention that here.

pkofod replied:
I could go either way. If it is not `one(T)` in other places similar to this change, I'd be happy to change it.

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +77,105 @@ function cos_kernel(x::DoubleFloat32)
+
+asin_q(t::Float32) = muladd(t, -7.0662963390f-01, 1.0f0)
+
+@inline function asin_kernel(t::Float64, x)
```
second argument `x::Float64`

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +77,105 @@ function cos_kernel(x::DoubleFloat32)
+        return flipsign(pi/4 - (p-q), x)
+    end
+end
+@inline function asin_kernel(t::Float32, x)
```
second argument `x::Float64`

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +77,105 @@ function cos_kernel(x::DoubleFloat32)
+ASIN_X_MIN_THRESHOLD(::Type{Float32}) = 2.0f0^-12
+ASIN_X_MIN_THRESHOLD(::Type{Float64}) = sqrt(eps(Float64))
+
+asin_p(t::Float64) = t*@horner(t,
```
alignment here looks a bit weird. Maybe write these as:
```julia
asin_p(t::Float64) =
    t * @horner(t,
    1.66666666666666657415e-01,
    -3.25565818622400915405e-01,
    2.01212532134862925881e-01,
    -4.00555345006794114027e-02,
    7.91534994289814532176e-04,
    3.47933107596021167570e-05)
```
?

musm commented on base/special/trig.jl:
```diff
@@ -76,6 +77,105 @@ function cos_kernel(x::DoubleFloat32)
+    # else 1/2 <= |x| < 1
+    t = (one(T) - absx)/2
+    return asin_kernel(t, x)
+end
```
BTW this function is inserted in a odd location in this file, breaking the other trig `sin`, `cos functions  organization.

Perhaps move it down to line : https://github.com/pkofod/julia/blob/e1767161572228843bfe4fda00d588250b3c06ff/base/special/trig.jl#L412

Or even it's own file (like the `exp` and `exp10` functions) and how openlibm has it organized ?

pkofod replied:
I don't know if "JuliaLang" prefers it to be here or in a separate file. I don't have an opinion myself, but I do get your point about moving it down.

musm commented on base/fastmath.jl:
```diff
@@ -292,6 +292,7 @@ atan2_fast(x::Float32, y::Float32) =
 atan2_fast(x::Float64, y::Float64) =
     ccall(("atan2",libm), Float64, (Float64,Float64), x, y)
 
+asin_fast(x::Union{Float32, Float64}) = asin(x)
```
Use `x::FloatTypes` (already defined above)

musm commented on base/special/trig.jl:
```diff
@@ -80,6 +81,111 @@ end
+    # Special cases:
+    #    if |x|>1, throw DomainError
+    absx = abs(x)
+    if absx >= T(1.0)) # |x|>= 1
```
extra `)`  (see also L185)

pkofod commented:
Cleaned up the commits so it can be merged without squashing should we want the added tests in place before the changes, though you can squash all you want.

Happy to take further comments should there be any.

musm commented:
Thanks @pkofod  this is LGMT. The only thing I would change is the `throw(DomainError(..))` and move that inline. In my tests there was no timing difference (on 0.6), and it would be preferable to avoid the extra function if it is unnecessary  (I can't benchmark this on master, yet but I suspect the timings should be same, otherwise all the `throw(Domain...` exception would need modification, I think that should not be the case).

KristofferC commented:
>  otherwise all the `throw(Domain...` exceptions would need modificatio

https://github.com/JuliaLang/julia/blob/0df3f7dc0280525349ae23e3ac4fc96621befc50/base/math.jl#L30

etc

musm commented:
@KristofferC  that function is used in more than one location, so I suspect that is not due to performance but to reduce code.

KristofferC commented:
@musm There is a reason it is `@noinline`.

musm commented:
@KristofferC I think you meant to refer to @pkofod .  @pkofod I was referring to having the errors  thrown within the function and not to define a separate function outside to do so, as @KristofferC and I were originally discussing (see comments at  https://github.com/JuliaLang/julia/pull/23088/files#r132786571).

KristofferC commented:
No I didn't and the function should not(!) be moved inline. I'm timing on a 15 days old master and get 8.5 ns with the function not inlined and ~10ns with it inlined.
The reason you don't notice a difference on 0.6 is probably because the DomainError with arguments doesn't exist there.

The start of the function without inlining:

```
julia> @code_llvm asin(0.5f0)

; Function asin
; Location: REPL[16]
define float @julia_asin_62167(float) #0 {
top:
; Location: REPL[16]:27
  %1 = call float @llvm.fabs.f32(float %0)
; Location: REPL[16]:28
  %2 = fcmp ult float %1, 1.000000e+00
  br i1 %2, label %L14, label %if
```

The start of the function with inlining:

```
julia> @code_llvm asin(0.5f0)

; Function asin
; Location: REPL[12]
define float @julia_asin_62034(float) #0 {
top:
  %1 = alloca [2 x %jl_value_t addrspace(10)*], align 8
  %gcframe7 = alloca [3 x %jl_value_t addrspace(10)*], align 8
  %gcframe7.sub = getelementptr inbounds [3 x %jl_value_t addrspace(10)*], [3 x %jl_value_t addrspace(10)*]* %gcframe7, i64 0, i64 0
  %2 = getelementptr inbounds [3 x %jl_value_t addrspace(10)*], [3 x %jl_value_t addrspace(10)*]* %gcframe7, i64 0, i64 1
  %3 = bitcast %jl_value_t addrspace(10)** %2 to i8*
  call void @llvm.memset.p0i8.i32(i8* %3, i8 0, i32 16, i32 8, i1 false)
  %ptls_i8 = call i8* asm "movq %fs:0, $0;\0Aaddq $$-10928, $0", "=r,~{dirflag},~{fpsr},~{flags}"() #5
  %4 = bitcast [3 x %jl_value_t addrspace(10)*]* %gcframe7 to i64*
  store i64 2, i64* %4, align 8
  %5 = getelementptr [3 x %jl_value_t addrspace(10)*], [3 x %jl_value_t addrspace(10)*]* %gcframe7, i64 0, i64 1
  %6 = bitcast i8* %ptls_i8 to i64*
  %7 = load i64, i64* %6, align 8
  %8 = bitcast %jl_value_t addrspace(10)** %5 to i64*
  store i64 %7, i64* %8, align 8
  %9 = bitcast i8* %ptls_i8 to %jl_value_t addrspace(10)***
  store %jl_value_t addrspace(10)** %gcframe7.sub, %jl_value_t addrspace(10)*** %9, align 8
; Location: REPL[12]:27
  %10 = call float @llvm.fabs.f32(float %0)
; Location: REPL[12]:28
  %11 = fcmp ult float %10, 1.000000e+00
  br i1 %11, label %L14, label %if
```

musm commented:
Just because the output from code_llvm   is longer doesn't  necessarily mean it will be slower

I checkout out the commit before the BenchmarkTools macro broke on master and here are my findings:
https://gist.github.com/musm/7b51c83e56bfc1a4de58df23be3db3f0

Conclusion:
There may be  up to a ~2 nanosecond slow down  in the minimum time benchmarks using a non-inline throw statement.

KristofferC commented:
Are you on windows? Your timer resolution looks bad (integral nanoseconds). Anyway, we are talking about like 20% of the total execution time of these functions so at this point it seems that you should strongly agree with me that the function should not be inlined? 

> Just because the output from code_llvm is longer doesn't necessarily mean it will be slower

I know, which is why I also benchmarked it... This is also code that is only additional in one of the functions, it's not like there's an extra SIMD loop or something there. Anyway, since we both have benchmarked a significant time difference and that the generated code is more bloated for no win, I suggest we stop discussing this and leave things as is.

musm commented:
I added benchmarks for a different machine https://gist.github.com/musm/7b51c83e56bfc1a4de58df23be3db3f0#gistcomment-2177534 
results look similar (i.e. less than 2 ns diff)

I agree there is a  performance advantage up to 20% ( it would be even faster to just throw a `NaN` for this and other cases, but that is of course not on the table 😉.)

I suspect this pattern of not  manually non-inlining the calls  will become very common so it would be better to  have the compiler  do this for me automatically. Hopefully future improvements will make this a non-issue. 

Sorry if I dragged this out but I think it's important to discuss because there are  other places where this crops up, e.g.:
https://github.com/JuliaLang/julia/blob/master/base/math.jl#L453
https://github.com/JuliaLang/julia/blob/master/base/math.jl#L612
See also the package https://github.com/JuliaMath/SpecialFunctions.jl where all of them are not manually `non-inlined` . So if the performance gap is up to 20%, shouldn't they all be  a manual no-inline (which would be super ugly and annoying to do)

Anyways, this would be better to discuss in another issue/forum and is also independent of this PR.

pkofod commented:
> @pkofod I was referring to having the errors thrown within the function and not to define a separate function outside to do so, as @KristofferC and I were originally discussing (see comments at https://github.com/JuliaLang/julia/pull/23088/files#r132786571).

Yes, yes, I knew that. I just meant that I was pretty sure that Kristoffer's were real (I tried it myself at the time) so I didn't want to remove the `noinline`ing before a benchmark showed it to be wrong. 2ns is quite a lot at this scale!

> ( it would be even faster to just throw a NaN for this and other cases, but that is of course not on the table :wink:.)

I agree, but I'm just following "tradition" :)

pkofod commented:
I'm sort of surprised at the performance improvements here, since there weren't really any in #23283 and the two sets of methods are essentially doing the same amount of work. Could we rerun these?

fredrikekre commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23088#issuecomment-323273000) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/e51450d265563696cede4c86a9264a480ec35fd1/172b17c_vs_a945af3/report.md). cc @ararslan

pkofod commented:
Interesting... Oh well, I guess that more of a `acos` PR thing then, I'll have to look at the code it generates compared to this one.

Unless someone has any comments or suggestions, I'd say this is good to go. `@musm` and Simon gave LGTM, and I addressed all comments I think. Thanks for all the comments and suggestions.

pkofod committed:
Change asin_p and asin_q for arc_p and arc_q for future use in acos.
